java-squid/src/main/java/org/sonar/java/cfg/CFG.java

```diff
@@ -0,0 +1,691 @@
+
+  public List<Block> blocks() {
+    return Lists.reverse(blocks);
+  }
+
+  public static class Block {
+    public final int id;
+    final List<Tree> elements = new ArrayList<>();
+    public List<Block> successors = new ArrayList<>();
+    public List<Block> predecessors = new ArrayList<>();
+    public Tree terminator;
+
+    public Block(int id) {
+      this.id = id;
+    }
+
+    public List<Tree> elements() {
+      return Lists.reverse(elements);
+    }
+  }
+
+  private CFG(BlockTree tree) {
+    exitBlock = createBlock();
+    currentBlock = createBlock(exitBlock);
+    for (StatementTree statementTree : Lists.reverse(tree.body())) {
+      build(statementTree);
+    }
+
+    for (Block b : gotos) {
+      assert b.successors.isEmpty();
+      Tree s = b.terminator;
+      assert s != null;
+      String label;
+      if(s instanceof BreakStatementTree) {
+        label = ((BreakStatementTree) s).label().name();
+      } else {
+        label = ((ContinueStatementTree) s).label().name();
+      }
+      Block target = labels.get(label);
+      if (target == null) {
+        throw new IllegalStateException("Undeclared label: " + label);
+      }
+      b.successors.add(target);
+    }
+
+    for (Block b : blocks) {
+      for (Block successor : b.successors) {
+        successor.predecessors.add(b);
+      }
+    }
+  }
+
+  private Block createBlock(Block successor) {
+    Block result = createBlock();
+    result.successors.add(successor);
+    return result;
+  }
+
+  private Block createBlock() {
+    Block result = new Block(blocks.size());
+    blocks.add(result);
+    return result;
+  }
+
+  public static CFG build(MethodTree tree) {
+    Preconditions.checkArgument(tree.block() != null, "Cannot build CFG for method with no body.");
+    return new CFG(tree.block());
+  }
+
+  private void build(List<? extends Tree> trees) {
+    for (Tree tree : Lists.reverse(trees)) {
+      build(tree);
+    }
+  }
+
+  private void build(Tree tree) {
+    switch (tree.kind()) {
+      case BLOCK:
+        for (StatementTree statementTree : Lists.reverse(((BlockTree) tree).body())) {
+          build(statementTree);
+        }
+        break;
+      case RETURN_STATEMENT: {
+        ReturnStatementTree s = (ReturnStatementTree) tree;
+        currentBlock = createUnconditionalJump(s, exitBlock);
+        ExpressionTree expression = s.expression();
+        if (expression != null) {
+          build(expression);
+        }
+        break;
+      }
+      case EXPRESSION_STATEMENT:
+        build(((ExpressionStatementTree) tree).expression());
+        break;
+      case METHOD_INVOCATION:
+        MethodInvocationTree mit = (MethodInvocationTree) tree;
+        currentBlock.elements.add(mit);
+        build(mit.methodSelect());
+        for (ExpressionTree arg : Lists.reverse(mit.arguments())) {
+          build(arg);
+        }
+        break;
+      case IF_STATEMENT: {
+        IfStatementTree ifStatementTree = (IfStatementTree) tree;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this line start at column 11. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AIndentationCheck)